src/merkle_tree/constraints.rs

```diff
@@ -101,75 +45,250 @@ where
+        // Thus `left_hash` is sibling if leaf_position_bit is 1, and leaf if bit is 0
+
+        let left_hash = self
+            .leaf_position_bit
+            .select(leaf_sibling_hash, &claimed_leaf_hash)?
+            .to_bytes()?;
+        let right_hash = self
+            .leaf_position_bit
+            .select(&claimed_leaf_hash, leaf_sibling_hash)?
+            .to_bytes()?;
+
+        let mut curr_hash =
+            TwoToOneH::evaluate_both(two_to_one_hash_parameter, &left_hash, &right_hash)?;
+        // To traverse up a MT, we iterate over the path from bottom to top (in reverse)
+
+        let path_rev: Vec<_> = self.path.iter().rev().collect();
+
+        let auth_path_rev: Vec<_> = self.auth_path.iter().rev().collect();
+
+        for level in 0..auth_path_rev.len() {
+            // At any given bit, the bit being 0 indicates our currently hashed value is the left,
+            // and the bit being 1 indicates our currently hashed value is on the right.
+            // Thus `left_hash` is the sibling if bit is 1, and it's the computed hash if bit is 0
+            let bit = path_rev[level];
+            let sibling = auth_path_rev[level];
+
+            let left_hash = bit.select(sibling, &curr_hash)?;
+            let right_hash = bit.select(&curr_hash, sibling)?;
+
+            curr_hash = TwoToOneH::evaluate_both(
+                two_to_one_hash_parameter,
+                &left_hash.to_bytes()?,
+                &right_hash.to_bytes()?,
+            )?;
+        }
+
+        Ok(curr_hash)
+    }
+
+    /// Given a leaf, calculate the hash of the merkle tree
+    /// along the path and check if the given root is correct.
+    ///
+    /// Constraints will not be satisfied if root is incorrect.
+    pub fn check_membership(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+    ) -> Result<(), SynthesisError> {
+        self.conditionally_check_membership(
+            leaf_hash_parameter,
+            two_to_one_hash_parameter,
+            root,
+            leaf,
+            &Boolean::constant(true),
+        )
+    }
+
+    /// Check membership if `should_enforce` is True.
+    pub fn conditionally_check_membership(
+        &self,
+        leaf_hash_parameter: &LeafH::ParametersVar,
+        two_to_one_hash_parameter: &TwoToOneH::ParametersVar,
+        root: &TwoToOneH::OutputVar,
+        leaf: &impl ToBytesGadget<ConstraintF>,
+        should_enforce: &Boolean<ConstraintF>,
+    ) -> Result<(), SynthesisError> {
+        let expected_root =
+            self.calculate_root(leaf_hash_parameter, two_to_one_hash_parameter, leaf)?;
+        // enforce `expected_root` is equal to the given root
+        expected_root.conditional_enforce_equal(root, should_enforce);
+        Ok(())
+    }
```

Comment:
In other places, we've moved away from `(conditionally)_check_*` which adds a constraint to enforce membership with a `verify_*` method that just returns a bit. Then the caller can decide whether they want to do `bit.is_eq(TRUE/FALSE)`.

In this case, that would mean returning `root_hash.is_eq(expected_root)?`